src/number_theory/sum_two_squares.lean

```diff
@@ -5,27 +5,227 @@ Authors: Chris Hughes
-  norm_num
+  rwa [principal_ideal_ring.irreducible_iff_prime, prime_iff_mod_four_eq_three_of_nat_prime p],
+end
+
+end Fermat
+
+/-!
+### Generalities on sums of two squares
+-/
+
+section general
+
+/-- The set of sums of two squares is closed under multiplication in any commutative ring.
+See also `sq_add_sq_mul_sq_add_sq`. -/
+lemma sq_add_sq_mul {R} [comm_ring R] {a b x y u v : R} (ha : a = x ^ 2 + y ^ 2)
+  (hb : b = u ^ 2 + v ^ 2) : ∃ r s : R, a * b = r ^ 2 + s ^ 2 :=
+⟨x * u - y * v, x * v + y * u, by {rw [ha, hb], ring}⟩
+
+/-- The set of natural numbers that are sums of two squares is closed under multiplication. -/
+lemma nat.sq_add_sq_mul {a b x y u v : ℕ} (ha : a = x ^ 2 + y ^ 2) (hb : b = u ^ 2 + v ^ 2) :
+  ∃ r s : ℕ, a * b = r ^ 2 + s ^ 2 :=
+begin
+  zify at ha hb ⊢,
+  obtain ⟨r, s, h⟩ := sq_add_sq_mul ha hb,
+  refine ⟨r.nat_abs, s.nat_abs, _⟩,
+  simpa only [int.coe_nat_abs, sq_abs],
+end
+
+end general
+
+/-!
+### Results on when -1 is a square modulo a natural number
+-/
+
+section neg_one_square
+
+/-- If `-1` is a square modulo `n` and `m` divides `n`, then `-1` is also a square modulo `m`. -/
+-- This could be formulated for a general integer `a` in place of `-1`,
+-- but it would not directly specialize to `-1`,
+-- because `((-1 : ℤ) : zmod n)` is not the same as `(-1 : zmod n)`.
+lemma zmod.is_square_cast_neg_one {m n : ℕ} (hd : m ∣ n) (hs : is_square (-1 : zmod n)) :
+  is_square (-1 : zmod m) :=
+begin
+  let f : zmod n →+* zmod m := zmod.cast_hom hd _,
+  rw [← ring_hom.map_one f, ← ring_hom.map_neg],
+  exact hs.map f,
 end
+
+/-- If `-1` is a square modulo coprime natural numbers `m` and `n`, then `-1` is also
+a square modulo `m*n`. -/
+-- A similar comment applies here.
+lemma zmod.is_square_neg_one_mul {m n : ℕ} (hc : m.coprime n) (hm : is_square (-1 : zmod m))
+  (hn : is_square (-1 : zmod n)) : is_square (-1 : zmod (m * n)) :=
+begin
+  have : is_square (-1 : (zmod m) × (zmod n)),
+  { rw show (-1 : (zmod m) × (zmod n)) = ((-1 : zmod m), (-1 : zmod n)), from rfl,
+    obtain ⟨x, hx⟩ := hm,
+    obtain ⟨y, hy⟩ := hn,
+    rw [hx, hy],
+    exact ⟨(x, y), rfl⟩, },
+  simpa only [ring_equiv.map_neg_one] using this.map (zmod.chinese_remainder hc).symm,
+end
+
+/-- If a prime `p` divides `n` such that `-1` is a square modulo `n`, then `p % 4 ≠ 3`. -/
+lemma nat.prime.mod_four_ne_three_of_dvd_is_square_neg_one {p n : ℕ} (hpp: p.prime) (hp : p ∣ n)
+  (hs : is_square (-1 : zmod n)) : p % 4 ≠ 3 :=
+begin
+  obtain ⟨y, h⟩ := zmod.is_square_cast_neg_one hp hs,
+  rw [← sq, eq_comm, show (-1 : zmod p) = -1 ^ 2, from by ring] at h,
+  haveI : fact p.prime := ⟨hpp⟩,
+  exact zmod.mod_four_ne_three_of_sq_eq_neg_sq' one_ne_zero h,
+end
+
+/-- If `n` is a squarefree natural number, then `-1` is a square modulo `n` if and only if
+`n` is not divisible by a prime `q` such that `q % 4 = 3`. -/
+lemma zmod.is_square_neg_one_iff {n : ℕ} (hn : squarefree n) :
+  is_square (-1 : zmod n) ↔ ∀ q : ℕ, q.prime → q ∣ n → q % 4 ≠ 3 :=
```

Comment:
I think it would be nice to add the following variant:
```lean
lemma zmod.is_square_neg_one_iff' {n : ℕ} (hn : squarefree n) :
  is_square (-1 : zmod n) ↔ ∀ q : ℕ, q ∣ n → q % 4 ≠ 3 :=
begin
  rw zmod.is_square_neg_one_iff hn,
  refine ⟨λ h q, unique_factorization_monoid.induction_on_prime q _ _ _, λ h q hq₁ hq₂, h q hq₂⟩,
  { rintros hq₀, rw zero_dvd_iff at hq₀, simp, },
  { simp only [nat.is_unit_iff], rintros - rfl -, simp, },
  { intros q' p hq' hp h₁ h₂,
    replace h₁ : q' % 4 ≠ 3 := h₁ (dvd_of_mul_left_dvd h₂),
    replace h₂ : p % 4 ≠ 3 := h p (nat.prime_iff.mpr hp) (dvd_of_mul_right_dvd h₂),
    -- etc.
   sorry, },
end
```